File: src/Data_Structure/Stack/Problems/MaximumBinaryTree/MBT.cpp
#include <bits/stdc++.h> 
using namespace std;

class Solution {

private: 
    int findMaxIndex(vector<int> &nums, int l, int r)
    {
        int maxI = l;

        for (int i = l; i<=r ; i++)
        {
            if(nums[i] >= nums[maxI])
                maxI = i;               
        }
        return maxI;
    }
    
    TreeNode* constructMaxSubtree(vector<int>& nums, int l, int r)//l, r are inclusive
    {
        if(l>r) return NULL;//not empty when l==r
        int maxI = findMaxIndex(nums, l, r); // this is max index so this nums will be root        
        TreeNode* root = new TreeNode(nums[maxI]);
        //Root is constructed now make children
        root->left = constructMaxSubtree(nums,l,maxI-1);
        root->right = constructMaxSubtree(nums, maxI+1,r);
        return root;
    }
    
public:
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        if(nums.size() == 0) return NULL;
        if(nums.size() == 1) return new TreeNode(nums[0]);
        TreeNode *maxTreeRoot;
        maxTreeRoot = constructMaxSubtree(nums,0,(nums.size()-1));
        return maxTreeRoot;
    }
};

// Time complexity : O(n^2)
// Space complexity : O(n)

File: src/Data_Structure/Stack/Problems/MaximumBinaryTree/MBT.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

package Data_Structure.Stack.Problems.MaximumBinaryTree;
import java.util.*;

public class MBT {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if (nums.length == 0) {
          return null;
        }
        TreeNode root = helper(nums, 0, nums.length - 1);
        return root;
      }
      
      private TreeNode helper(int[] nums, int start, int end) {
        if (start > end) {
          return null;
        }
        int maxIdx = getMaxIdx(nums, start, end);
        TreeNode root = new TreeNode(nums[maxIdx]);
        root.left = helper(nums, start, maxIdx - 1);
        root.right = helper(nums, maxIdx + 1, end);
        return root;
      }
      
      private int getMaxIdx(int[] nums, int start, int end) {
        int maxVal = Integer.MIN_VALUE;
        int maxIdx = -1;
        while (start <= end) {
          if (nums[start] > maxVal) {
            maxVal = nums[start];
            maxIdx = start;
          }
          start++;
        }
        return maxIdx;
      }
}

// Time Complexity : O(n^2)
// Space Complexity : O(h)